### PR TITLE
Reference live ILLink analyzer

### PR DESCRIPTION
**Pull request by sbomer**
In https://github.com/dotnet/runtime/pull/91233 I inadvertently turned off the trim analyzer for libraries, because the analyzer  bits don't flow to the consuming project with the ILLink.Tasks ProjectReference. This adds a separate ProjectReference to use the live analyzer. This addresses https://github.com/dotnet/runtime/issues/88901 for the ILLink analyzers.

Using the live analyzer uncovered some bugs that are fixed in this change:
- Invalid assert for field initializers with `throw` statements
- Invalid cast to `IMethodSymbol` for the ContainingSymbol of a local. Locals can occur in field initializers, when created from out params.
- Wrong warning code produced for assignment to annotated property in an initializer. This was being treated as a call to the setter instead of an assignment to the underlying field:
   `Type AnnotatedPropertyWithSetter { get; set; } = GetUnknown ();`
   (IIRC I hit this one when investigating some related behavior, not via a new warning produced when analyzing runtime)
- Invalid assert for delegate creation where the delegate is created from anything other than a method (for example, a field that has an Action or Func type).

### CODE REVIEW

File: eng/liveILLink.targets [hunk 1]
```diff
@@ -39,6 +39,12 @@
         <SetTargetFramework Condition="'$(MSBuildRuntimeType)' == 'Core'">TargetFramework=$(NetCoreAppToolCurrent)</SetTargetFramework>
         <SetTargetFramework Condition="'$(MSBuildRuntimeType)' != 'Core'">TargetFramework=$(NetFrameworkToolCurrent)</SetTargetFramework>
     </ProjectReference>
+
+    <!-- Need to reference the analyzer project separately, because there's no easy way to get it as a transitive reference of ILLink.Tasks.csproj -->
+    <ProjectReference Include="$(_ILLinkTasksSourceDir)..\ILLink.RoslynAnalyzer\ILLink.RoslynAnalyzer.csproj"
+                      ReferenceOutputAssembly="false"
+                      OutputItemType="Analyzer"
```
Comment (ViktorHofer) [inline eng/liveILLink.targets:46]: Please also set PrivateAssets and Private same as the above ProjectReference sets.

File: src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs [hunk 2]
```diff
@@ -190,6 +195,11 @@ TValue ProcessSingleTargetAssignment (IOperation targetOperation, ISimpleAssignm
 						// For now, just don't warn. https://github.com/dotnet/linker/issues/2731
```
Comment (vitek-karas) [inline src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs:195]:
```suggestion
						// For now, just don't warn. https://github.com/dotnet/runtime/issues/93277
```

I moved the issue to runtime - and we should look into fixing this.

**Review by vitek-karas — APPROVED** [review]

File: eng/liveILLink.targets (hunk 1 above)
Comment (ViktorHofer) [reply eng/liveILLink.targets:46]: Those would probably also benefit from `Private="false"`. That hinders msbuild from copying the source generator into the lib/app output which is unnecessary work. Not 100% sure about PrivateAssets though. Maybe that metadata is by default true when `ReferenceOutputAssembly=false`.

**Review by vitek-karas — APPROVED** [review]

File: eng/liveILLink.targets (hunk 1 above)
Comment (ViktorHofer) [reply eng/liveILLink.targets:47]: OK, I just verified that `PrivateAssets` isn't required here. The ILLink.RoslynAnalyzer.csproj dependency doesn't show up as a package dependency with your changes, i.e. for System.Text.Json:

image

**Review by ViktorHofer — APPROVED** [review]
Also verified that this change doesn't require updating our solution files again in src/libraries.

**Pull request merged**